dev/tools/coqdev.el

```diff
@@ -0,0 +1,102 @@
+;;; coqdev.el --- Emacs helpers for Coq development  -*- lexical-binding:t -*-
+
+;; Copyright (C) 2018 The Coq Development Team
+
+;; Maintainer: coqdev@inria.fr
+
+;;; Commentary:
+
+;; Helpers to set compilation commands, proof general variables, etc
+;; for Coq development
+
+;; You can disable individual features without editing this file by
+;; using `remove-hook', for instance
+;; (remove-hook 'hack-local-variables-hook #'coqdev-setup-compile-command)
+
+;;; Installation:
+
+;; To use this, with coqdev.el located at /path/to/coqdev.el, add the
+;; following to your init:
+
+;; (add-to-list 'load-path "/path/to/coqdev/")
+;; (require 'coqdev)
+
+;; If you load this file from a git repository, checking out an old
+;; commit will make it disappear and cause errors for your Emacs
+;; startup. To ignore those errors use (require 'coqdev nil t). If you
+;; check out a malicious commit Emacs startup would allow it to run
+;; arbitrary code, to avoid this you can copy coqdev.el to any
+;; location and adjust the load path accordingly (of course if you run
+;; ./configure to compile Coq it is already too late).
+
+;;; Code:
+
+(require 'subr-x)
+
+(defun coqdev-default-directory ()
+  "Return the Coq repository containing `default-directory'."
+  (when-let ((dir (locate-dominating-file default-directory "META.coq")))
+    (expand-file-name dir)))
+
+(defun coqdev-setup-compile-command ()
+  "Setup `compilate-command' for Coq development."
+  (when-let ((dir (coqdev-default-directory)))
+      (setq-local compile-command (concat "make -C " dir))))
+(add-hook 'hack-local-variables-hook #'coqdev-setup-compile-command)
+
+(defvar camldebug-command-name) ; from camldebug.el (caml package)
+(defvar ocamldebug-command-name) ; from ocamldebug.el (tuareg package)
+(defun coqdev-setup-camldebug ()
+  "Setup ocamldebug for Coq development.
+
+Specifically `camldebug-command-name' and `ocamldebug-command-name'."
+  (when-let ((dir (coqdev-default-directory)))
+    (setq-local camldebug-command-name (concat dir "dev/ocamldebug-coq"))
+    (setq-local ocamldebug-command-name (concat dir "dev/ocamldebug-coq"))))
+(add-hook 'hack-local-variables-hook #'coqdev-setup-camldebug)
+
+(defun coqdev-setup-tags ()
+  "Setup `tags-file-name' for Coq development."
+  (when-let ((dir (coqdev-default-directory)))
+      (setq-local tags-file-name (concat dir "TAGS"))))
+(add-hook 'hack-local-variables-hook #'coqdev-setup-tags)
+
+(defvar coq-prog-args)
+(defvar coq-prog-name)
+
+;; Lets us detect whether there are file local variables
+;; even though PG sets it with `setq' when there's a _Coqproject.
+;; Also makes sense generally, so might make it into PG someday.
+(make-variable-buffer-local 'coq-prog-args)
+(setq-default coq-prog-args nil)
```

Comment:
What is this for?

Reply:
For a couple files ((C)Morphisms.v and some test suite files) there are file local arguments which we don't want to override, but we don't want to false positive detect the presence of file local arguments when PG set them from _CoqProject in another file (eg the files in Init/, or some 3rd party file).

Reply:
Yup, I understand the buffer-local part; but why do you set it to nil?

Reply:
In case PG modified it before us.